src/Symfony/Component/Messenger/Transport/Semaphore/Connection.php

```diff
@@ -0,0 +1,178 @@
+        $dsn = preg_replace('#^(semaphore):///#', '$1://localhost/', $dsn);
+
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Semaphore Messenger DSN "%s" is invalid.', $dsn));
+        }
+
+        $parsedQuery = [];
+        parse_str($parsedUrl['query'] ?? '', $parsedQuery);
+
+        $queueOptions = array_replace_recursive([
+            'path' => $parsedUrl['path'],
+        ], $options, $parsedQuery);
+
+        return new self($queueOptions);
+    }
+
+    public function getConfiguration(): array
+    {
+        return $this->configuration;
+    }
+
+    /**
+     * Send message on semaphore queue.
+     *
+     * @param SemaphoreStamp $semaphoreStamp
+     *
+     * @throws SemaphoreException
+     */
+    public function send(string $body, array $headers = [], int $delay = 0, SemaphoreStamp $semaphoreStamp = null): void
+    {
+        $messageType = null !== $semaphoreStamp ? $semaphoreStamp->getType() : ($this->configuration['message_type'] ?? 1);
+        $message = json_encode(['body' => $body, 'headers' => $headers]);
+
+        if ($this->configuration['message_max_size'] < \strlen($message)) {
+            throw new SemaphoreException(sprintf('The semaphore message is too long to be sent, the maximum size accepted is %s bytes.', $this->configuration['message_max_size']));
+        }
+
+        if (false === msg_send($this->getQueue(), $messageType, $message, false, false, $errorCode)) {
+            throw new SemaphoreException(sprintf('Semaphore sending message failed with error code : "%s".', $errorCode));
+        }
+    }
+
+    /**
+     * Waits and gets a message from the configured semaphore queue.
+     *
+     * @throws SemaphoreException
+     */
+    public function get(): ?SemaphoreEnvelope
+    {
+        if (true === msg_receive($this->getQueue(), $this->configuration['message_type'] ?? 0, $messageType, $this->configuration['message_max_size'], $message, false, MSG_IPC_NOWAIT, $errorCode)) {
+            $message = json_decode($message, true);
+
+            return new SemaphoreEnvelope($messageType, $message['body'], $message['headers']);
+        }
+
+        if (MSG_ENOMSG !== $errorCode) {
+            throw new SemaphoreException(sprintf('Semaphore receiving message failed with error code : "%s".', $errorCode));
+        }
+
+        return null;
+    }
+
+    public function getMessageCount(): int
+    {
+        $stat = msg_stat_queue($this->getQueue());
+
+        return $stat['msg_qnum'] ?? 0;
+    }
+
+    public function close(): void
+    {
+        msg_remove_queue($this->getQueue());
+    }
+
+    public function setup(): void
+    {
+        $key = ftok($this->configuration['path'], substr($this->configuration['project'], 0, 1));
+        $this->queue = msg_get_queue($key);
+        $this->setup = true;
+    }
+
+    private function shouldSetup(): bool
```

Comment:
this method could be replaced by a `auto_setup` property. Initialized in constructor see Redis for example